### PR TITLE
Lookup NTG on Leaf Grid View

### PR DESCRIPTION
**Pull request by aritorto**
To support LGRs, the lookup of NTG has been updated, namely, for a leaf grid view cell, the ntg value is looked up via its parent cell (if existent), or its equivalent cell in level 0.  

Needs OPM/opm-grid#689 .

### CODE REVIEW

**Comment by blattms** [comment]
This seems to be done differently than for other field properties. Maybe we should handle it consistently and make sure that the arrays are always of the size of elements on the leaf grid view?

E.g. in #4951 we made sure that we have arrays that live on the leaf grid view using `LookUpData::assignFieldPropsIntOnLeaf`, etc. . In a similar way there could be a `getFieldPropsDoubleOnLeaf` that returns the same as `eclState_.fieldProps().get_double(property_name)` without LGR and with LGR does project the values to an array on the leaf grid. The we would need to change only one line [ecltransmissibility_impl.hh#L158](https://github.com/OPM/opm-simulators/blob/ea9bc396570b38948b170b0c78c0ffb65974d617/ebos/ecltransmissibility_impl.hh#L158).

**Comment by blattms** [comment]
There are some warnings about shadowed variables: https://ci.opm-project.org/job/opm-grid-PR-builder/478/gcc/

**Comment by aritorto** [comment]
Indeed, thanks for noticing. We can use the existing method in one line.

**Comment by aritorto** [comment]
jenkins build this please

File: ebos/ecltransmissibility_impl.hh [hunk 1]
```diff
@@ -469,8 +474,8 @@ update(bool global, const std::function<unsigned int(unsigned int)>& map, const
                                                         axisCentroids),
                                         porosity_[outsideElemIdx]);
 
-                applyNtg_(halfDiffusivity1, insideFaceIdx, elemIdx, ntg);
-                applyNtg_(halfDiffusivity2, outsideFaceIdx, outsideElemIdx, ntg);
+                applyNtg_(halfDiffusivity1, insideFaceIdx, coarseElemIdx, ntg);
+                applyNtg_(halfDiffusivity2, outsideFaceIdx, coarseOutsideElemIdx, ntg);
```
Comment (blattms) [inline ebos/ecltransmissibility_impl.hh:478]: Does this still work after the change above? I thought it should be elemIdx outsideElemIdx as before.

File: ebos/ecltransmissibility_impl.hh [hunk 2]
```diff
@@ -1038,7 +1043,7 @@ applyNncMultreg_(const std::unordered_map<std::size_t,int>& cartesianToCompresse
     const auto& inputNNC = this->eclState_.getInputNNC();
     const auto& transMult = this->eclState_.getTransMult();
 
-    auto compressedIdx = [&cartesianToCompressed](const std::size_t globIdx)
+    auto compressedIdx = [&cartesianToCompressed, this](const std::size_t globIdx)
```
Comment (blattms) [inline ebos/ecltransmissibility_impl.hh:1046]: "this" does not seem to be used. Please don't capture it.

File: ebos/ecltransmissibility_impl.hh [hunk 3]
```diff
@@ -495,7 +500,7 @@ update(bool global, const std::function<unsigned int(unsigned int)>& map, const
     // Loop over all elements (global grid) and store Cartesian index
     for (const auto& elem : elements(grid_.leafGridView())) {
         int elemIdx = elemMapper.index(elem);
-        int cartElemIdx = cartMapper_.cartesianIndex(elemIdx);
+        int cartElemIdx =  this->lookUpCartesianData_.template getFieldPropCartesianIdx<Grid>(elemIdx);
```
Comment (blattms) [inline ebos/ecltransmissibility_impl.hh:503]: Can you remind me what the difference between the calls is?
I was under the impression that both should do the same. In that case we should probably use the old version.

**Review by blattms — CHANGES_REQUESTED** [review]
I think there is a problem with the changes now. ntg_ has entries for each element of the leaf grid, but we use the indices of the parent to access its elements.

File: ebos/ecltransmissibility_impl.hh (hunk 2 above)
Comment (aritorto) [reply ebos/ecltransmissibility_impl.hh:1046]: Thanks, it was a left over of an old-wrong change. I'll remove it

File: ebos/ecltransmissibility_impl.hh (hunk 3 above)
Comment (aritorto) [reply ebos/ecltransmissibility_impl.hh:503]: They do the same with and without LGRs, when all the field properties are assumed to be given in the level 0/coarse grid. The difference is when field properties for LGRs are given in their respected refined grids (levels>0). Then, the lookUpCartesianData_ method will get the "local/level" Cartesian index of the leaf grid view cell, on the level this cell was born, and this is not necessarily level 0.

With the current assumptions, we can of course leave the old line of code, and go back to change it when refined field properties will be supported.

File: ebos/ecltransmissibility_impl.hh (hunk 1 above)
Comment (aritorto) [reply ebos/ecltransmissibility_impl.hh:478]: Indeed, thanks for noticing it!

File: ebos/ecltransmissibility_impl.hh (hunk 3 above)
Comment (aritorto) [reply ebos/ecltransmissibility_impl.hh]: it's back to the old version

**Comment by aritorto** [comment]
jenkins build this serial please

**Review by blattms — APPROVED** [review]

**Comment by blattms** [comment]
benchmark please

**Comment by ytelses** [comment]
Benchmark result overview: 




| Test    | Configuration                                     | Relative |
| ------- | ------------------------------------------------- | -------- |
| opm-git | OPM Benchmark: drogon - Threads: 1                | 0.998    |
| opm-git | OPM Benchmark: drogon - Threads: 8                | 1.006    |
| opm-git | OPM Benchmark: punqs3 - Threads: 1                | 0.995    |
| opm-git | OPM Benchmark: punqs3 - Threads: 8                | 1.006    |
| opm-git | OPM Benchmark: smeaheia - Threads: 1              | 1.01     |
| opm-git | OPM Benchmark: smeaheia - Threads: 8              | 1.002    |
| opm-git | OPM Benchmark: spe10_model_1 - Threads: 1         | 1.012    |
| opm-git | OPM Benchmark: spe10_model_1 - Threads: 8         | 1.007    |
| opm-git | OPM Benchmark: flow_mpi_extra - Threads: 1        | 0.994    |
| opm-git | OPM Benchmark: flow_mpi_extra - Threads: 8        | 1.002    |
| opm-git | OPM Benchmark: flow_mpi_norne - Threads: 1        | 0.999    |
| opm-git | OPM Benchmark: flow_mpi_norne - Threads: 8        | 1.002    |
| opm-git | OPM Benchmark: flow_mpi_norne_4c_msw - Threads: 1 | 0.999    |
| opm-git | OPM Benchmark: flow_mpi_norne_4c_msw - Threads: 8 | 1.003    |
* Speed-up = Total time master / Total time pull request. Above 1.0 is an improvement. *



View result details @ https://www.ytelses.com/opm/?page=result&id=2289

**Comment by blattms** [comment]
Cool, no performance penalty with this.

**Pull request merged**